fdb-record-layer-spatial/src/main/java/com/apple/foundationdb/record/spatial/geophile/GeophileIndexMaintainer.java

```diff
@@ -27,28 +27,56 @@
 import com.apple.foundationdb.record.RecordCursor;
 import com.apple.foundationdb.record.ScanProperties;
 import com.apple.foundationdb.record.TupleRange;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.apple.foundationdb.record.metadata.Index;
+import com.apple.foundationdb.record.metadata.expressions.KeyExpression;
+import com.apple.foundationdb.record.metadata.expressions.KeyExpressionWithChild;
+import com.apple.foundationdb.record.metadata.expressions.KeyWithValueExpression;
 import com.apple.foundationdb.record.provider.foundationdb.IndexMaintainerState;
 import com.apple.foundationdb.record.provider.foundationdb.indexes.StandardIndexMaintainer;
 import com.geophile.z.Space;
 
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+import java.util.List;
 
 /**
  * The index maintainer class for (geo-)spatial indexes.
  *
  */
 @API(API.Status.EXPERIMENTAL)
 public class GeophileIndexMaintainer extends StandardIndexMaintainer {
-    // TODO: In order to make this parametric on index options, there needs to be a way to pass it down
-    //  to the evaluation function.
-    public static final Space SPACE_LAT_LON = GeophileSpatial.createLatLonSpace();
     @Nonnull
     private final Space space;
 
     public GeophileIndexMaintainer(IndexMaintainerState state) {
         super(state);
-        this.space = SPACE_LAT_LON;
+        this.space = getSpatialFunction(state.index).getSpace();
+    }
+
+    @Nonnull
+    // If the bottom-right child is GeophileSpatialFunctionKeyExpression, return it. Else error.
+    static GeophileSpatialFunctionKeyExpression getSpatialFunction(@Nonnull Index index) {
+        KeyExpression rootKey = index.getRootExpression();
+        if (rootKey instanceof KeyWithValueExpression) {
+            rootKey = ((KeyWithValueExpression)rootKey).getKeyExpression();
+        }
+        final List<KeyExpression> components = rootKey.normalizeKeyForPositions();
+        final KeyExpression rightComponent = components.get(components.size() - 1);
+        KeyExpression bottomComponent = rightComponent;
+        while (true) {
+            if (bottomComponent instanceof GeophileSpatialFunctionKeyExpression) {
+                return (GeophileSpatialFunctionKeyExpression)bottomComponent;
+            }
+            if (bottomComponent instanceof KeyExpressionWithChild) {
+                bottomComponent = ((KeyExpressionWithChild)bottomComponent).getChild();
```

Comment:
Yeah, they are supposed to go in the value. They "work" in the key for that same purpose, but they can't be used there for any additional comparisons because the Z-order lookup needs to do one or more range scans on that column.

Reply:
Okay, I see. That makes sense.